source/rust_verify/src/singular.rs

```diff
@@ -339,67 +336,66 @@ pub fn check_singular_valid(
         panic!("internal error: integer_ring")
     };
 
-    let arc_ens: &air::ast::Stmt = stmts.last().unwrap();
-    let ens = match &**arc_ens {
-        air::ast::StmtX::Assert(error, exp) => (exp.clone(), error.clone()),
-        _ => {
-            panic!("internal error");
-        }
-    };
-
-    let mut vars: Vec<Ident> = vec![];
+    let mut vars: Vec<String> = vec![];
     for d in &**decl {
         if let air::ast::DeclX::Var(name, _typ) = &**d {
-            vars.push(name.clone());
+            let mut v2: String = name.to_string();
+            v2.push('@');
+            vars.push(sanitize_var_name(v2));
         }
     }
 
-    let mut reqs = vec![];
-    for idx in 0..(stmts.len() - 1) {
-        // last element is ensures clause
-        let stm = &stmts[idx];
-        match &**stm {
-            air::ast::StmtX::Assert(error, exp) => {
-                reqs.push((exp.clone(), error.clone()));
-            }
-            _ => {
-                panic!("internal error");
-            }
-        };
-    }
-
-    let query = match singular_printer(&vars, &reqs, &ens) {
-        Ok(query_string) => query_string,
-        Err(err) => return ValidityResult::Invalid(None, err),
-    };
-
-    air::singular_manager::log_singular(context, &query, func_span);
-
-    let quit_string = format!("{};", QUIT_SINGULAR);
-    let query = format!("{} {}", query, quit_string);
-
+    let mut encoder = SingularEncoder::new(vars);
     let singular_manager = SingularManager::new();
     let mut singular_process = singular_manager.launch();
-    let res = singular_process.send_commands(query.as_bytes().to_vec());
-    if (res.len() == 1) && (res[0] == "0") {
-        ValidityResult::Valid
-    } else if res[0].contains("?") {
-        ValidityResult::UnexpectedOutput(String::from(format!(
-            "{} \ngenerated singular query: {}",
-            res[0].as_str(),
-            query
-        )))
-    } else {
-        ValidityResult::Invalid(
-            None,
-            air::messages::error(
-                format!(
-                    "postcondition not satisfied: Ensures polynomial failed to be reduced to zero, reduced polynomial is {}\n generated singular query: {} ",
```

Comment:
My understanding is that with this change we are now only emitting "Failed to translate this expression into a singular query" as the failure message. That may not be ideal, as it doesn't indicate to the user if what they're trying to prove is not true, or if they wrote an invalid query.

Can "Failed to translate this expression into a singular query" indicate that the requires/ensures is invalid? Or is it always the equivalent of an "assertion violation" message?

Reply:
I have made some updates and the error messages are different depending on: 1. invalid input 2. internal encoding error 3. failed proof

Reply:
Great, thanks!